ipld/read.go

```diff
@@ -18,190 +18,162 @@ import (
-	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, flattened, codec, tree.Constructor)
-}
 
-// uniqueRandNumbers generates count unique random numbers with a max of max
-func uniqueRandNumbers(count, max int) []uint32 {
-	if count > max {
-		panic(fmt.Sprintf("cannot create %d unique samples from a max of %d", count, max))
-	}
-	samples := make(map[uint32]struct{}, count)
-	for i := 0; i < count; {
-		// nolint:gosec // G404: Use of weak random number generator
-		sample := uint32(rand.Intn(max))
-		if _, has := samples[sample]; has {
-			continue
-		}
-		samples[sample] = struct{}{}
-		i++
-	}
-	out := make([]uint32, count)
-	counter := 0
-	for s := range samples {
-		out[counter] = s
-		counter++
-	}
-	return out
+	return extended, batchAdder.Commit()
 }
 
-type index struct {
-	row uint32
-	col uint32
+type quadrant struct {
+	isLeftSubtree bool
+	from          int
+	chunk         [][]byte //roots
 }
 
-type indexedShare struct {
-	data []byte
-	index
-}
+func pickRandomQuadrant(roots [][]byte) *quadrant {
+	edsWidth := len(roots)
+	// get the random number from 1 to 4.
+	q := rand.Intn(4) + 1
+	quadrant := &quadrant{chunk: make([][]byte, edsWidth/2)}
+	// quadrants 1 and 3 corresponds to left subtree,
+	// 2 and 4 to the right subtree
+	/* | 1 | 2 |
+	   | 3 | 4 |
+	*/
+	// choose sub-tree
+	if q%2 == 1 {
+		quadrant.isLeftSubtree = true
+	}
 
-// shareCounter is a thread safe tallying mechanism for share retrieval
-type shareCounter struct {
-	// all shares
-	shares map[index][]byte
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
+	// define range of shares for sampling
+	if q > 2 {
+		quadrant.from = edsWidth / 2
+		copy(quadrant.chunk, roots[edsWidth/2:])
+	} else {
+		copy(quadrant.chunk, roots[:edsWidth/2])
+	}
 
-	shareChan chan indexedShare
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
+	return quadrant
 }
 
-func newshareCounter(parentCtx context.Context, edsWidth uint32) *shareCounter {
-	ctx, cancel := context.WithCancel(parentCtx)
+// fillQuadrant fetches 1/4 of shares for the given root
+func fillQuadrant(
+	ctx context.Context,
+	roots *quadrant, dag ipld.NodeGetter,
+	isRow bool,
+	dataSquare [][]byte,
+) error {
+	errGroup, ctx := errgroup.WithContext(ctx)
+	fetcher := func(i int) {
+		errGroup.Go(func() error {
+			rootHash, err := plugin.CidFromNamespacedSha256(roots.chunk[i])
+			if err != nil {
+				return err
+			}
+			subtreeRootHash, err := GetSubtreeLeaves(ctx, rootHash, dag, roots.isLeftSubtree)
+			if err != nil {
+				return err
+			}
 
-	// calculate the min number of shares needed to repair the square
-	minSharesNeeded := edsWidth * edsWidth / 4
+			leaves, err := GetLeaves(ctx, dag, subtreeRootHash, uint32(len(roots.chunk)))
```

Comment:
Should be the size of roots divided by two

Reply:
done